gapis/api/vulkan/api/queued_command_tracking.api

```diff
@@ -108,13 +108,76 @@ sub void execPendingCommands(VkQueue queue) {
+                    widthInBlocks := roundUpTo(level.Width, elementAndTexelBlockSize.TexelBlockSize.Width)
+                    heightInBlocks := roundUpTo(level.Height, elementAndTexelBlockSize.TexelBlockSize.Height)
+                    elementSize := switch (aspectBit) {
+                      case VK_IMAGE_ASPECT_COLOR_BIT:
+                        elementAndTexelBlockSize.ElementSize
+                      case VK_IMAGE_ASPECT_DEPTH_BIT:
+                        depthElementSize
+                      case VK_IMAGE_ASPECT_STENCIL_BIT:
+                        // stencil element is always 1 byte wide
+                        as!u32(1)
+                    }
+                    size := widthInBlocks * heightInBlocks * level.Depth * elementSize
+                    level.Data = make!u8(size)
+                  }
+                }
+              }
+            }
           }
 
           for _ , img , binds in cmd.SparseBinds.ImageBinds {
             if (!img in Images) { vkErrorInvalidImage(img) }
             for _ , _ , bind in binds.SparseImageMemoryBinds {
               addSparseImageMemoryBinding(img, bind)
             }
+            image := Images[img]
+            // Creates the shadow image data memory here if has not been done so.
+            // This is exactly the same to vkBindImageMemory except for linear
+            // and preinitialized images, we also create individual pool here.
+            // TODO: Redirect to bound memories for the data of linear preinitialized
+            // images.
+            for _, _, aspectBit in unpackImageAspectFlags(image.ImageAspect).Bits {
+              aspect := image.Aspects[aspectBit]
+              for lay in (0 .. image.Info.ArrayLayers) {
+                for lev in (0 .. image.Info.MipLevels) {
+                  level := aspect.Layers[lay].Levels[lev]
+                  if len(level.Data) == 0 {
+                    elementAndTexelBlockSize := getElementAndTexelBlockSize(image.Info.Format)
+                    depthElementSize := getDepthElementSize(image.Info.Format, false)
+                    // Roundup the width and height in the number of blocks.
+                    widthInBlocks := roundUpTo(level.Width, elementAndTexelBlockSize.TexelBlockSize.Width)
+                    heightInBlocks := roundUpTo(level.Height, elementAndTexelBlockSize.TexelBlockSize.Height)
+                    elementSize := switch (aspectBit) {
+                      case VK_IMAGE_ASPECT_COLOR_BIT:
+                        elementAndTexelBlockSize.ElementSize
+                      case VK_IMAGE_ASPECT_DEPTH_BIT:
+                        depthElementSize
+                      case VK_IMAGE_ASPECT_STENCIL_BIT:
+                        // stencil element is always 1 byte wide
+                        as!u32(1)
+                    }
+                    size := widthInBlocks * heightInBlocks * level.Depth * elementSize
+                    level.Data = make!u8(size)
```

Comment:
We will have to see what this looks like if/when we actually switch our pools over to actual byte slices. This will likely not be what we want.

Reply:
I think we need to 2 steps
1) Get the texture data from replay side, so that nothing rely on the image level data pool any more
2) Move to use the device memory data pool.

Or as you said, we can move to the pool first, for the cases that the calculated size is smaller than the required size.